src/permission-set.js

```diff
@@ -16,523 +16,809 @@
+     * @property authorizations
+     * @type {Object}
+     */
+    this.authorizations = {}
+    /**
+     * The URL of the corresponding ACL resource, at which these permissions will
+     * be saved.
+     * @property aclUrl
+     * @type {String}
+     */
+    this.aclUrl = aclUrl
+    /**
+     * Optional request host (used by checkOrigin())
+     * @property host
+     * @type {String}
+     */
+    this.host = options.host
+    /**
+     * Initialize the agents / groups / resources indexes.
+     * @property index
+     * @type {Object}
+     */
+    this.index = {
+      'agents': {},
+      'groups': {}  // Also includes Public permissions
+    }
+    /**
+     * RDF Library (optionally injected)
+     * @property rdf
+     * @type {RDF}
+     */
+    this.rdf = options.rdf
+    /**
+     * Whether this permission set is for a 'container' or a 'resource'.
+     * Determines whether or not the inherit/'acl:default' attribute is set on
+     * all its Authorizations.
+     * @property resourceType
+     * @type {String}
+     */
+    this.resourceType = isContainer ? CONTAINER : RESOURCE
+    /**
+     * The URL of the resource for which these permissions apply.
+     * @property resourceUrl
+     * @type {String}
+     */
+    this.resourceUrl = resourceUrl
+    /**
+     * Should this permission set enforce "strict origin" policy?
+     * (If true, uses `options.origin` parameter)
+     * @property strictOrigin
+     * @type {Boolean}
+     */
+    this.strictOrigin = options.strictOrigin
+    /**
+     * Contents of the request's `Origin:` header.
+     * (used only if `strictOrigin` parameter is set to true)
+     * @property origin
+     * @type {String}
+     */
+    this.origin = options.origin
+    /**
+     * Solid REST client (optionally injected), used by save() and clear().
+     * @type {SolidWebClient}
+     */
+    this.webClient = options.webClient
+
+    // Init the functions for deriving an ACL url for a given resource
+    this.aclUrlFor = options.aclUrlFor
+      ? options.aclUrlFor
+      : PermissionSet.defaultAclUrlFor
+    this.aclUrlFor.bind(this)
+    this.isAcl = options.isAcl ? options.isAcl : PermissionSet.defaultIsAcl
+    this.isAcl.bind(this)
+
+    // Optionally initialize from a given parsed graph
+    if (options.graph) {
+      this.initFromGraph(options.graph)
+    }
+  }
+
   /**
-   * Hashmap of all Authorizations in this permission set, keyed by a hashed
-   * combination of an agent's/group's webId and the resourceUrl.
-   * @property authorizations
-   * @type {Object}
+   * Returns the corresponding ACL uri, for a given resource.
+   * This is the default `aclUrlFor()` method that's used by
+   * PermissionSet instances, unless it's overridden in options.
+   * @method defaultAclUrlFor
+   * @param resourceUri {String}
+   * @return {String} ACL uri
    */
-  this.authorizations = {}
+  static defaultAclUrlFor (resourceUri) {
+    if (this.isAcl(resourceUri)) {
```

Comment:
So if I understand correctly you're allowing clients to chuck a `defaultAclUrlFor` method on to the class dynamically, but you want a default one in case none is specified in the constructor.

a) I don't understand or agree with the assumed use case. You're providing this dynamic behavior, but as far as I can tell no client would actually make use of it. The web access control spec states "Clients MUST NOT assume that the location of an ACL resource can be deterministically derived from a document's URL," so a method called `defaultAclUrlFor` which takes a resourceUri and deterministically derives the ACL resource from that document URL is misaligned with our spec.

b) My feeling is that abusing the semantics of class static methods is the wrong way to satisfy that use case.